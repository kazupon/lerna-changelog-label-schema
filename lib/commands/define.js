"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const utils_1 = require("../utils");
const debug_1 = require("debug");
const debug = debug_1.debug('lerna-changelog-label-schema:commands:define');
const define = async (options) => {
    const { preset, cwd } = options;
    debug(`preset= ${preset}, cwd=${cwd}`);
    if (!(preset === 'default' || preset === 'full')) {
        return false;
    }
    if (!cwd) {
        return false;
    }
    const schema = await Promise.resolve().then(() => __importStar(require(`../schema/${preset}`)));
    const labels = schema.labels;
    debug('labels', labels);
    let pkgPath = '';
    let pkgJSON = {};
    try {
        pkgPath = path.resolve(cwd, './package.json');
        pkgJSON = require(pkgPath);
        debug('pkgJSON', pkgJSON);
    }
    catch (e) {
        return false;
    }
    // TODO: shoudl be typed changelogCaptions
    const changelogCaptions = labels.reduce((captions, label) => {
        captions[`Type: ${label.type}`] = `${label.emoji} ${label.caption}`;
        return captions;
    }, {});
    debug('changelogCaptions', changelogCaptions);
    const newPkgJSON = applyToPacakge(pkgJSON, changelogCaptions);
    debug('newPkgJson', newPkgJSON);
    utils_1.writeJSON(pkgPath, newPkgJSON);
    console.log(`apply lerna-changelog label '${preset}' schema to package.json`);
    return true;
};
function applyToPacakge(pkg, labels) {
    return Object.assign(pkg, { changelog: { labels } });
}
exports.default = define;
