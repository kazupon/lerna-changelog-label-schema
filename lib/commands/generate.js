"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const debug_1 = require("debug");
const debug = debug_1.debug('lerna-changelog-label-schema:commands:generate');
const generate = async (options) => {
    const { preset, output } = options;
    debug(`preset= ${preset}, output=${output}`);
    if (!(preset === 'default' || preset === 'full')) {
        return false;
    }
    const schema = await Promise.resolve().then(() => __importStar(require(`../schema/${preset}`)));
    const labels = schema.labels;
    const meta = schema.meta;
    debug('labels', labels);
    debug('meta', meta);
    const githubLabels = labels.map((label, index) => {
        const m = meta[index];
        if (m.type !== label.type) {
            throw new Error(`Invalid generating: label=${label.type}, meta=${m.type}, index=${index}`);
        }
        const ghLabel = {
            name: `Type: ${label.type}`,
            color: m.color
        };
        if (label.description) {
            ghLabel.description = label.description;
        }
        if (m.aliases) {
            ghLabel.aliases = m.aliases;
        }
        return ghLabel;
    });
    debug('githubLabels', githubLabels);
    let ret = true;
    if (output) {
        try {
            await utils_1.writeJSON(output, githubLabels);
        }
        catch (e) {
            console.error(e.message);
            ret = false;
        }
    }
    else {
        console.log(JSON.stringify(githubLabels, null, 2));
    }
    return ret;
};
exports.default = generate;
