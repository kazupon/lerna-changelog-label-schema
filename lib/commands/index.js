"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const define_1 = __importDefault(require("./define"));
exports.define = define_1.default;
const generate_1 = __importDefault(require("./generate"));
exports.generate = generate_1.default;
