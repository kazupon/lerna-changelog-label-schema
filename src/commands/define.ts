import * as path from 'path'
import { writeJSON } from '../utils'
import {
  LernaChangelogLabel,
  CommandHandler,
  CommandOption
} from '../../types'

import { debug as Debug } from 'debug'
const debug = Debug('lerna-changelog-label-schema:commands:define')

const define: CommandHandler = async (options: CommandOption) => {
  const { preset, cwd } = options
  debug(`preset= ${preset}, cwd=${cwd}`)

  if (!(preset === 'default' || preset === 'full')) {
    return false
  }

  if (!cwd) {
    return false
  }

  const schema = await import(`../schema/${preset}`)
  const labels = schema.labels as LernaChangelogLabel[]
  debug('labels', labels)

  let pkgPath = ''
  let pkgJSON = {} as JSON
  try {
    pkgPath = path.resolve(cwd as string, './package.json')
    pkgJSON = require(pkgPath) as JSON
    debug('pkgJSON', pkgJSON)
  } catch (e) {
    return false
  }

  // TODO: shoudl be typed changelogCaptions
  const changelogCaptions = labels.reduce((captions: any, label) => {
    captions[`Type: ${label.type}`] = `${label.emoji} ${label.caption}`
    return captions
  }, {})
  debug('changelogCaptions', changelogCaptions)

  const newPkgJSON = applyToPacakge(pkgJSON, changelogCaptions)
  debug('newPkgJson', newPkgJSON)
  writeJSON(pkgPath, newPkgJSON)
  console.log(`apply lerna-changelog label '${preset}' schema to package.json`)

  return true
}

function applyToPacakge (pkg: JSON, labels: JSON) {
  return Object.assign(pkg, { changelog: { labels }})
}

export default define
