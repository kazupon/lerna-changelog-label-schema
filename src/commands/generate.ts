import { writeJSON } from '../utils'
import {
  GitHubLabelMeta,
  LernaChangelogLabel,
  GitHubLabel,
  CommandHandler,
  CommandOption
} from '../../types'

import { debug as Debug } from 'debug'
const debug = Debug('lerna-changelog-label-schema:commands:generate')

const generate: CommandHandler = async (options: CommandOption) => {
  const { preset, output } = options
  debug(`preset= ${preset}, output=${output}`)

  if (!(preset === 'default' || preset === 'full')) {
    return false
  }

  const schema = await import(`../schema/${preset}`)
  const labels = schema.labels as LernaChangelogLabel[]
  const meta = schema.meta as GitHubLabelMeta[]
  debug('labels', labels)
  debug('meta', meta)

  const githubLabels = labels.map((label, index) => {
    const m = meta[index]
    if (m.type !== label.type) {
      throw new Error(`Invalid generating: label=${label.type}, meta=${m.type}, index=${index}`)
    }
    const ghLabel: GitHubLabel = {
      name: `Type: ${label.type}`,
      color: m.color
    }
    if (label.description) {
      ghLabel.description = label.description
    }
    if (m.aliases) {
      ghLabel.aliases = m.aliases
    }
    return ghLabel
  }) as GitHubLabel[]
  debug('githubLabels', githubLabels)

  let ret = true
  if (output) {
    try {
      await writeJSON(output as string, githubLabels)
    } catch (e) {
      console.error(e.message)
      ret = false
    }
  } else {
    console.log(JSON.stringify(githubLabels, null, 2))
  }

  return ret
}

export default generate
