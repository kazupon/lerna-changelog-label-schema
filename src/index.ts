import * as _default from './schema/default'
import * as full from './schema/full'

export const schema = {
  default: _default,
  full
}
