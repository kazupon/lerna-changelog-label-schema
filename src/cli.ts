import meow from 'meow'
import { define, generate } from './commands/index'

import { debug as Debug } from 'debug'
const debug = Debug('lerna-changelog-label-schema:cli')

export async function run (argv: string[] = []) {
  const options: meow.Options = {
    flags: {
      preset: {
        type: 'string',
        alias: 'p',
        default: 'default'
      },
      output: {
        type: 'string',
        alias: 'o'
      }
    }
  }

  if (argv.length > 0) {
    options.argv = argv
  }

  const cli = meow(`
  Usage
    $ lerna-changelog-label-schema <command> [options]
  
  Commands:
    define, d      define lerna-changelog labels in your package.json 
    generate, g    generate lerna-changelog labels for GitHub labels, output for stdout
    version, v     show the version

  Options:
    --preset, -p   Label schema preset option, for define and generate command. 'default' or 'full', default: 'default'
    --output, -o   Output option, for generate command. if specified, stdout is suppressed.
  
  Examples:
    $ lerna-changelog-label-schema define --preset=full
    $ lerna-changelog-label-schema generate --preset=default
  `, options)

  const command = cli.input.pop() || ''
  debug('command:', command)
  const { preset, output } = cli.flags
  debug('flags:', cli.flags)

  if (['define', 'd'].includes(command)) { // define command
    await define({ preset, cwd: process.cwd() })
  } else if (['generate', 'g'].includes(command)) { // generate command
    await generate({ preset, output })
  } else if (['version', 'v'].includes(command)) { // version command
    console.log(cli.pkg.version)
  } else { // help!
    console.log(cli.help)
  }

  return cli
}
