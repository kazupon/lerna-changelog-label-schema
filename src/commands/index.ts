import define from './define'
import generate from './generate'

export {
  define,
  generate
}
