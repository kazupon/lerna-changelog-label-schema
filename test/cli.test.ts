// -------
// mocking

const TARGET_PATH = '/path/to/project1'

let orgCwd // for process.cwd mock

// mock: ../src/commands
jest.mock('../src/commands', () => ({
  define: jest.fn(),
  generate: jest.fn()
}))
import * as commands from '../src/commands'

// -------------------
// setup/teadown

let spyLog
beforeEach(() => {
  orgCwd = process.cwd
  process.cwd = jest.fn(() => TARGET_PATH) // mock: process.cwd
  spyLog = jest.spyOn(global.console, 'log')
})

afterEach(() => {
  spyLog.mockRestore()
  jest.clearAllMocks()
  process.cwd = orgCwd
})

// -------------------
// tests

test('define command', async () => {
  // setup mocks
  const { define } = commands as jest.Mocked<typeof commands>
  define.mockImplementation(options => Promise.resolve(true))

  // run
  const { run } = await import('../src/cli')
  const cli = await run(['define', '--preset=full'])

  // verify
  expect(cli.flags).toMatchObject({ preset: 'full', p: 'full' })
  expect(define).toHaveBeenCalledWith({ preset: 'full', cwd: TARGET_PATH })
})

test('generate command', async () => {
  // setup mocks
  const { generate } = commands as jest.Mocked<typeof commands>
  generate.mockImplementation(options => Promise.resolve(true))

  // run
  const { run } = await import('../src/cli')
  const cli = await run(['g', '--output=/path/to/preset.json'])

  // verify
  expect(cli.flags).toMatchObject({ preset: 'default', p: 'default', output: '/path/to/preset.json' })
  expect(generate).toHaveBeenCalledWith({ preset: 'default', output: '/path/to/preset.json' })

})

test('version command', async () => {
  // run
  const { run } = await import('../src/cli')
  const cli = await run(['version'])

  // verify
  expect(spyLog).toHaveBeenCalledWith(cli.pkg.version)
})

test('show help', async () => {
  // run
  const { run } = await import('../src/cli')
  const cli = await run()

  // verify
  expect(spyLog.mock.calls[0][0]).toMatchSnapshot()
})