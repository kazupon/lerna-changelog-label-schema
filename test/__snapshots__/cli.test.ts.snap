// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show help 1`] = `
"
  Label schema definition for lerna-changelog

  Usage
    $ lerna-changelog-label-schema <command> [options]

  Commands:
    define, d      define lerna-changelog labels in your package.json 
    generate, g    generate lerna-changelog labels for GitHub labels, output for stdout
    version, v     show the version

  Options:
    --preset, -p   Label schema preset option, for define and generate command. 'default' or 'full', default: 'default'
    --output, -o   Output option, for generate command. if specified, stdout is suppressed.

  Examples:
    $ lerna-changelog-label-schema define --preset=full
    $ lerna-changelog-label-schema generate --preset=default

"
`;
